benches/nexmark/generator/auctions.rs

```diff
@@ -11,6 +11,55 @@ use std::{
 };
 
 impl<R: Rng> NexmarkGenerator<R> {
+    /// Return the last valid auction id (ignoring FIRST_AUCTION_ID). Will be
+    /// the current auction id if due to generate an auction.
+    fn last_base0_auction_id(&self, event_id: u64) -> u64 {
+        let mut epoch = event_id / self.config.nexmark_config.total_proportion() as u64;
+        let mut offset = event_id % self.config.nexmark_config.total_proportion() as u64;
+
+        if offset < self.config.nexmark_config.person_proportion as u64 {
+            // About to generate a person.
+            // Go back to the last auction in the last epoch.
+            epoch = match epoch.checked_sub(1) {
+                Some(e) => e,
+                None => return 0,
+            };
+            offset = self.config.nexmark_config.auction_proportion as u64 - 1;
+        } else if offset
+            >= (self.config.nexmark_config.person_proportion
+                + self.config.nexmark_config.auction_proportion) as u64
+        {
+            // About to generate a bid.
+            // Go back to the last auction generated in this epoch.
+            offset = self.config.nexmark_config.auction_proportion as u64 - 1;
+        } else {
+            // About to generate an auction.
+            offset -= self.config.nexmark_config.person_proportion as u64;
+        }
+        epoch * self.config.nexmark_config.auction_proportion as u64 + offset
+    }
+
+    /// Return a random auction id (base 0).
+    fn next_base0_auction_id(&mut self, next_event_id: u64) -> u64 {
+        // Choose a random auction for any of those which are likely to still be in
+        // flight, plus a few 'leads'.
+        // Note that ideally we'd track non-expired auctions exactly, but that state
+        // is difficult to split.
+        let min_auction = cmp::max(
+            match self
+                .last_base0_auction_id(next_event_id)
+                .checked_sub(self.config.nexmark_config.num_in_flight_auctions as u64)
+            {
+                Some(e) => e,
+                None => 0,
+            },
+            0,
+        );
```

Comment:
I'm reasonably sure this is just `.saturating_sub()`, although I'm not really sure what the `max()` is doing here
```suggestion
        let min_auction = self
                .last_base0_auction_id(next_event_id)
                .saturating_sub(self.config.nexmark_config.num_in_flight_auctions as u64);
```

Reply:
Oh - yes, it's amazing what Rust has built in to the std types compared to Java :) I think you're right, and I assume the `max()` with zero is only there because in the Java implementation it's a long (ie. signed) type. Thanks.